### DISCUSSION TITLE
Determining the integration rules in the kernel

### DISCUSSION THREAD

**Discussion by teddyz1999** [idaholab/moose#27252] [Q&A General]
Hi，
When solving an equation with multiple variables, where the orders of these variables are specified in the input file, and the equation contains a nonlinear term involving the multiplication of multiple variables, how does MOOSE determine the integration rules for this nonlinear term within the kernel, such as the order of integration and the number of integration points?

**Comment by GiudGiud** [comment]
Hello

The default quadrature is not selected based on the form of the equation. It is chosen to be a Gauss quadrature of order n to be able to integrate polynomial products (for example shape * test function). It's exact in 1D
https://mooseframework.inl.gov/syntax/Executioner/Quadrature/index.html

Guillaume

**Reply by teddyz1999** [reply]
After reading the link you provided, I'm still a bit confused. It says "we actually use a quadrature of order 2n+1, with n the order of the variables," which means if my variable's basis function is of the second order, then a quadrature rule is chosen to precisely satisfy 2*2+1=5, resulting in a 5th order polynomial. However, if I set different orders for several variables in the input file, for example, one variable is of the second order and another is of the third order, how is the quadrature determined?

**Reply by GiudGiud** [reply]
Hello

We use the maximum quadrature order required, and we use it for all variables
see SystemBase::getMinQuadratureOrder() for polling the quadrature orders from variables
and
FEProblemBase::createQRules for polling each system for the quadrature order they need

**Reply by teddyz1999** [reply]
So， If I want to manually specify the quadrature order instead of letting it be automatically determined in MOOSE, how should I do that? Can I simply call an interface in the kernel to set a specific quadrature order for different kernels?

**Reply by GiudGiud** [reply]
Hello

you cannot specify different quadratures on a per kernel basis without rewriting computeResidual (instead of computeQpResidual) and being very careful with the re initialization of material properties and variable values. 

We can do (spatial) block-dependent quadratures fairly easily inside the Executioner/Quadrature block

Guillaume
